repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java

```diff
@@ -370,13 +381,35 @@ private boolean isFilteredChildAssociation(QName childAssocType, String user)
 
     private EventInfo getEventInfo(String user)
     {
-        return new EventInfo().setTimestamp(ZonedDateTime.now())
+        return new EventInfo().setTimestamp(getCurrentTransactionTimestamp())
                               .setId(UUID.randomUUID().toString())
                               .setTxnId(AlfrescoTransactionSupport.getTransactionId())
                               .setPrincipal(user)
                               .setSource(URI.create("/" + descriptorService.getCurrentRepositoryDescriptor().getId()));
     }
 
+    private ZonedDateTime getCurrentTransactionTimestamp()
+    {
+        Long currentTransactionId = nodeDAO.getCurrentTransactionId(false);
```

Comment:
This method works relying on the execution of this method inside the same Thread that is executing the transaction. If that changes at same point, an alternative method should be adopted.

Reply:
Afaik the idea is to produce messages from the events (serialise the java object into text and etc) by a separate thread pool of workers. The events are still generated after the transaction commit stage.